src/tiled/stampbrush.cpp

```diff
@@ -380,13 +391,56 @@ void StampBrush::drawPreviewLayer(const QVector<QPoint> &list)
         QRegion paintedRegion;
         QVector<PaintOperation> operations;
         QHash<TileLayer *, QRegion> regionCache;
+        QHash<TileLayer *, TileLayer *> shiftedCopies;
 
         for (const QPoint &p : list) {
             const TileStampVariation variation = mStamp.randomVariation();
             mapDocument()->unifyTilesets(variation.map, mMissingTilesets);
 
             TileLayer *stamp = variation.tileLayer();
 
+            Map::StaggerAxis mapStaggerAxis = mapDocument()->map()->staggerAxis();
+            Map::StaggerIndex mapStaggerIndex = mapDocument()->map()->staggerIndex();
+            Map::StaggerIndex stampStaggerIndex = variation.map->staggerIndex();
+
+
+            //if staggered map, makes sure stamp stays the same
+            if (mapDocument()->map()->isStaggered()
+                    && ((mapStaggerAxis == Map::StaggerY)? stamp->height() > 1 : stamp->width() > 1)) {
+
+                if (mapStaggerAxis == Map::StaggerY) {
+                    bool topIsOdd = (p.y() - stamp->height() / 2) & 1;
+
+                    if ((stampStaggerIndex == mapStaggerIndex) == topIsOdd) {
+                        stamp = variation.tileLayer()->copy(variation.tileLayer()->bounds());
+                        shiftedCopies.insert(stamp, stamp);
```

Comment:
No, it's not easier with a bool flag, because you need a bool flag _for each variation_. This also means we're storing potentially multiple copies (one for each variation that ends up being used). Hence we need the QHash. You see if you've made a copy by doing:
```c++
TileLayer *shiftedStamp = shiftedCopies.value(stamp);
if (!shiftedStamp) {
    shiftedStamp = static_cast<TileLayer*>(stamp->clone());
    shiftedCopies.insert(stamp, shiftedStamp);
    // do the shifting
}
stamp = shiftedStamp;
```